proposals/p2687.md

```diff
@@ -0,0 +1,323 @@
+# Termination algorithm for impl selection
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/2687)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Non-type arguments](#non-type-arguments)
+    -   [Proof of termination](#proof-of-termination)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Measure complexity using type tree depth](#measure-complexity-using-type-tree-depth)
+    -   [Consider each type parameter in an `impl` declaration separately](#consider-each-type-parameter-in-an-impl-declaration-separately)
+    -   [Consider types in the interface being implemented as distinct](#consider-types-in-the-interface-being-implemented-as-distinct)
+    -   [Require some count to decrease](#require-some-count-to-decrease)
+
+<!-- tocstop -->
+
+## Abstract
+
+This proposal replaces the termination algorithm for `impl` selection. The
+previous algorithm relied on a recursion limit, which is counter to
+[our goal for predictability](/docs/design/generics/goals.md#predictability).
+
+The replacement is to terminate if any `impl` lookup performed while considering
+an `impl` declaration depends transitively on the same `impl` declaration with a
+"strict superset" of the types in the query.
+
+## Problem
+
+Consider this `impl` declaration:
+
+```
+interface I;
+impl forall [T:! type where Optional(.Self) is I] T as I;
+```
+
+A type like `i32` is a valid value of `T`, and so implements `I`, if
+`Optional(i32)` implements `I`. This `impl` declaration could also possibly be
+used to give an implementation of `Optional(i32)` for interface `I`, but only if
+there was an implementation of `Optional(Optional(i32))` for interface `I`. The
+job of the termination rule is to report an error instead of being caught in an
+infinite loop in this situation.
+
+Ideally, a termination rule would identify the loop in a minimal way. This has a
+few benefits, including reducing compile times and making error messages as
+short and understandable as they can be. One downside of the original recursion
+limit rule is its tendency to only detect a problem after the loop had been
+repeated many times. This is made worse if the recursion limit is large.
```

Comment:
```suggestion
Ideally, a termination rule would identify the loop in a minimal way. This has a
few benefits, including reducing compile times and making error messages as
short and understandable as possible. One downside of the original recursion
limit rule is its tendency to only detect a problem after the loop had been
started. This becomes worse if the recursion limit is large.
```

Reply:
I'm going to undo the "repeated many times" -> "started" edit since it changes the meaning in a way that misses the point I was trying to make.